aten/src/ATen/core/ATenDispatch.h

```diff
@@ -40,17 +42,47 @@ static inline TensorTypeId dispatchTypeId(TensorTypeSet ts) {
 
 }
 
+namespace detail {
+  struct MultiDispatchTensorTypeSet : IterArgs<MultiDispatchTensorTypeSet> {
+    TensorTypeSet ts;
+    void operator()(const at::Tensor& x) {
+      ts = ts | x.type_set();
+    }
+    void operator()(TensorOptions x) {
+      ts = ts | x.type_set();
+    }
+    void operator()(at::ArrayRef<at::Tensor> xs) {
+      for (const auto& x : xs) {
+        ts = ts | x.type_set();
+      }
+    }
+    template <typename T>
+    void operator()(const T& x) {
+      // do nothing
+    }
+  };
+
+  template <typename... Args>
+  TensorTypeSet multi_dispatch_tensor_type_set(Args&&... args) {
```

Comment:
Weird that this compiles without you moving arguments in. You should probably change the arguments to const& nevertheless.